### ISSUE TITLE
[investigation] Consider, if you will, the oauth library

### ISSUE THREAD

**Issue by tsmethurst** [superseriousbusiness/gotosocial#537]
Our oauth library is getting out of date.

We should investigate whether it's easier at this point to either:

1. Update to the upstream version -- https://github.com/go-oauth2/oauth2 -- and delete our fork.
2. Pull upstream changes into our fork and keep using that.
3. Use a different + more widely-used library as mentioned in https://github.com/superseriousbusiness/gotosocial/issues/197

This issue isn't for fixing this, since it's not desperately urgent, it's just for investigation + discussion.

**Comment by tsmethurst** [comment]
I did an investigation and here's what I found!

In terms of the differences between current upstream main and our fork, we're not missing anything important or anything that compromises security: https://github.com/superseriousbusiness/oauth2/compare/main...go-oauth2:master. As of the moment of making this comment, the main changes are being able to pass a context into `PasswordAuthorizationHandler`, a couple dependency updates, and a `PreRedirectErrorHandler` which we don't really need to use.

Given that these are only minor changes, there's currently no need to do either 1 or 2 imo.

I started investigating possibility 3. My findings so far are that https://github.com/ory/fosite looks like the better library, mostly because it's used in https://github.com/ory/hydra and can therefore be consider more hardened/tested. It also offers scope parsing and checking, which is ahead of what the currently library offers.

So I think we should, some time soon, take some time to try removing our current oauth2 implementation and putting ory/fosite in there instead, and then testing and making sure the refactor doesn't break anything, before we roll it out. Then when we've got that implemented to the same level of functionality as our current library, we can begin hooking in the extra features offered by fosite.
